src/ai/util/areaCodeFromZip.util.ts

```diff
@@ -0,0 +1,178 @@
+import { Injectable, Logger } from '@nestjs/common'
+import { differenceInYears, formatISO } from 'date-fns'
+import { S3Service } from 'src/vendors/aws/services/s3.service'
+import { AiService } from '../ai.service'
+
+const ZIP_TO_AREA_CODE_FILE = 'zip-to-area-code-mappings.json'
+const CACHE_EXPIRY_YEARS = 1
+
+type ZipToAreaCodeEntry = {
+  areaCodes: string[]
+  lastFetchedAt: string // ISO datestring
+}
+
+type ZipToAreaCodeMapping = Record<string, ZipToAreaCodeEntry>
+
+@Injectable()
+export class AreaCodeFromZipService {
+  private readonly logger = new Logger(AreaCodeFromZipService.name)
+
+  constructor(
+    private readonly s3Service: S3Service,
+    private readonly aiService: AiService,
+  ) { }
+
+  /**
+   * Gets area codes for a given zip code.
+   * First checks S3 cache, then falls back to OpenAI if not found.
+   * Automatically saves new mappings to S3.
+   */
+  async getAreaCodeFromZip(zipCode: string): Promise<string[] | null> {
+    if (!zipCode) {
+      return null
+    }
+
+    const normalizedZipCode = zipCode.split('-')[0].trim()
+
+    // Check cache first
+    const cachedEntry = await this.getAreaCodesFromS3(normalizedZipCode)
+    if (cachedEntry) {
+      const isExpired = this.isCacheExpired(cachedEntry.lastFetchedAt)
+      if (!isExpired) {
+        this.logger.debug(
+          `Found valid cached area codes for ${normalizedZipCode}`,
+          cachedEntry.areaCodes,
+        )
+        return cachedEntry.areaCodes
+      }
+      this.logger.debug(
+        `Cached area codes for ${normalizedZipCode} expired, re-fetching from OpenAI`,
+      )
+    }
+
+    this.logger.debug(
+      `Area codes for zip code ${normalizedZipCode} not found in cache, calling OpenAI`,
+    )
+    const areaCodes = await this.getAreaCodesFromOpenAI(normalizedZipCode)
+
+    if (areaCodes && areaCodes.length > 0) {
+      await this.saveAreaCodesToS3(normalizedZipCode, areaCodes)
+      return areaCodes
+    }
+
+    return null
+  }
+
+  private isCacheExpired(lastFetchedAt: string): boolean {
+    const lastFetched = new Date(lastFetchedAt)
+    const now = new Date()
+
+    return differenceInYears(now, lastFetched) >= CACHE_EXPIRY_YEARS
+  }
+
+  private async getAreaCodesFromS3(
+    zipCode: string,
+  ): Promise<ZipToAreaCodeEntry | null> {
+    try {
+      const bucket = process.env.ZIP_TO_AREA_CODE_BUCKET
+      if (!bucket) {
+        throw new Error(
+          'ZIP_TO_AREA_CODE_BUCKET environment variable is required',
+        )
+      }
```

Comment:
This should follow the pattern we have elsewhere in the project of checking for env vars at the top of the file, outside of the class, and throwing the error to kill the process when there's a missing/mis-configured env var detected.

Reply:
ok

Reply:
https://github.com/thegoodparty/gp-api/pull/860#discussion_r2547723514